### PR TITLE
adding methods for sendMany for batching transactions

### PR DESCRIPTION
**Pull request by sbhatnagar30**
sendMany will allow us to batch transactions and send as one tx, first use case will be to batch settlements

### CODE REVIEW

File: lib/btc/BtcRpc.js [hunk 1]
```diff
@@ -45,10 +45,33 @@ class BtcRpc {
     return this.asyncCall('cmdlineUnlock', [time]);
   }
 
+  async sendMany({ obj, options }) {
+    for (let tx in obj) {
+      obj[tx] /= 1e8;
+    }
+    let array = [];
```
Comment (micahriggan) [inline lib/btc/BtcRpc.js]: const array = ["", obj]

File: lib/btc/BtcRpc.js [hunk 2]
```diff
@@ -45,10 +45,33 @@ class BtcRpc {
...
+
   async sendToAddress({ address, amount }) {
     return this.asyncCall('sendToAddress', [address, amount / 1e8]);
   }
 
+  async unlockAndSendManyBatched({ obj, passphrase, options }) {
+    if (passphrase === undefined) {
+      passphrase = await passwordPromptAsync('> ');
+    }
+    await this.walletUnlock({ passphrase, time: 10 });
```
Comment (micahriggan) [inline lib/btc/BtcRpc.js]: 10 is too low for many wallets

Comment (sbhatnagar30) [reply lib/btc/BtcRpc.js]: What time do you suggest? I was following `unlockAndSendToAddress` as an example

Comment (micahriggan) [reply lib/btc/BtcRpc.js]: That one is likely broken as well.

Comment (sbhatnagar30) [reply lib/btc/BtcRpc.js]: in `unlockAndSendToAddressMany` time is defaulted to 10800, I have done this in my method and also `unlockAndSendToAddress`, my guess is this is what John wanted. Will confirm when he gets back

Comment (sbhatnagar30) [reply lib/btc/BtcRpc.js]: John confirmed, 10800 seconds is fine (3 hours)

File: lib/btc/BtcRpc.js [hunk 3]
```diff
@@ -45,15 +45,36 @@ class BtcRpc {
...
+      array.push(options);
+    }
+    return this.asyncCall('sendMany', array);
+  }
+
   async sendToAddress({ address, amount }) {
     return this.asyncCall('sendToAddress', [address, amount / 1e8]);
   }
 
+  async unlockAndSendManyBatched({ obj, passphrase, options, time = 10800 }) {
```
Comment (sbhatnagar30) [inline lib/btc/BtcRpc.js]: So I actually want to change this to accept an array of already batched inputs, and then iterate and call sendMany on those inputs

File: lib/btc/BtcRpc.js [hunk 4]
```diff
@@ -45,15 +45,67 @@ class BtcRpc {
...
+    }
+    return this.asyncCall('sendMany', array);
+  }
+
   async sendToAddress({ address, amount }) {
     return this.asyncCall('sendToAddress', [address, amount / 1e8]);
   }
 
+  async unlockAndSendManyBatched({ batchArray, passphrase, options, time = 10800 }) {
+    function RpcException(message) {
```
Comment (micahriggan) [inline lib/btc/BtcRpc.js]: pull this out

File: tests/bch.js [hunk 5]
```diff
@@ -167,4 +167,51 @@ describe('BCH Tests', function() {
     assert(isValid === false);
   });
 
+  it('should be able to send a bached transaction', async() => {
+    let address1 = 'bchreg:qq2lqjaeut5ppjkx9339htfed8enx7hmugk37ytwqy';
+    let amount1 = 10000;
+    let address2 = 'bchreg:qq6n0n37mut4353m9k2zm5nh0pejk7vh7u77tan544';
+    let amount2 = 20000;
+    const obj = {};
```
Comment (micahriggan) [inline tests/bch.js]: naming

File: tests/bch.js [hunk 6]
```diff
@@ -167,4 +167,51 @@ describe('BCH Tests', function() {
...
+
+    let address3 = 'bchreg:qz07vf90w70s8d0pfx9qygxxlpgr2vwz65d53p22cr';
+    let amount3 = '30000';
+    let address4 = 'bchreg:qzp2lmc7m49du2n55qmyattncf404vmgnq8gr53aj7';
+    let amount4 = '40000';
+    const obj2 = {};
+    obj2[address3] = amount3;
+    obj2[address4] = amount4;
+
+    const obj = [obj1, obj2];
```
Comment (micahriggan) [inline tests/bch.js]: naming

File: lib/btc/BtcRpc.js [hunk 7]
```diff
@@ -45,15 +45,67 @@ class BtcRpc {
...
+      passphrase = await passwordPromptAsync('> ');
+    }
+    await this.walletUnlock({ passphrase, time: time });
+    this.emitter.emit('unlocked', time);
+    let txSuccessResults = [];
+    let txFailureResults = [];
+    let failureBool;
+    for (let batch of batchArray) {
+      const emitData = {
+        output: batch
```
Comment (micahriggan) [inline lib/btc/BtcRpc.js]: can we add txid as a key here and set it to empty string? That way we'll no from looking at it that we're going to set that later

File: lib/btc/BtcRpc.js (hunk 4 above)
Comment (sbhatnagar30) [reply lib/btc/BtcRpc.js]: So I throw a custom error object instead now copied here:

```
    if (failureBool) {
      throw {
        name: 'Batched Transaction Failure',
        message: 'At least one of many requests Failed',
        data: txFailureResults
      };
    }
```

File: lib/btc/BtcRpc.js [hunk 8]
```diff
@@ -53,65 +64,81 @@ class BtcRpc {
...
-        amount: request.amount,
-        id: request.id
+    let payToArrayResult = [];
+    while(payToArray.length) {
+      let currentValue = 0;
+      let currentOutputs = 0;
+      let batch = {};
+      let tempArray = [];
+      while(currentValue < maxValue && currentOutputs < maxOutputs) {
+        let tx = payToArray.shift();
```
Comment (micahriggan) [inline lib/btc/BtcRpc.js]: Does this have side-effects?

File: lib/btc/BtcRpc.js [hunk 9]
```diff
@@ -53,65 +64,81 @@ class BtcRpc {
...
-        errorObject.success[i] = txid;
+        let txid = await this.sendMany({ batch });
+        emitData.data = batch;
         emitData.txid = txid;
-        this.emitter.emit('success', emitData);
-        result.push(txid);
-      } catch (error) {
-        errorObject.failure[i] = error;
+        let voutCounter = 1;
+        for (let tx of tempArray) {
```
Comment (micahriggan) [inline lib/btc/BtcRpc.js]: I don't really understand this block decorating tx with txid and vout, where does that get used?

Comment (sbhatnagar30) [reply lib/btc/BtcRpc.js]: it gets used on the settlement side, `tempArray` is basically each settlement in a single batch (and it preserves duplicates unlike the batch) and I am just adding the `txid` and `vout` that the settlement is a part of and concating that to a main array which I return back. The main array will have a 1 to 1 correlation to the `payToArray` passed in, just with the `txid` and `vout` also attached

File: lib/btc/BtcRpc.js [hunk 10]
```diff
@@ -53,65 +64,81 @@ class BtcRpc {
...
+      let currentValue = 0;
+      let currentOutputs = 0;
+      let batch = {};
+      let tempArray = [];
+      while(currentValue < maxValue && currentOutputs < maxOutputs) {
+        let tx = payToArray.shift();
+        tempArray.push(tx);
+        let address = tx.address;
+        let amount = tx.amount;
+        if (address in batch) {
```
Comment (micahriggan) [inline lib/btc/BtcRpc.js]: I would do something like 

```
if(batch[address] === undefined) {
  batch[address] = 0;
}
batch[address] += amount
```

The way you're doing it here will result in currentValue being increased more than it should for dupes.

IE [{"abc": 1}, {"abc": 1}] would result in currentValue being 1, then current value being 3

File: lib/btc/BtcRpc.js (hunk 8 above)
Comment (sbhatnagar30) [reply lib/btc/BtcRpc.js]: This only returns and removes the first element, no side-effects that I am aware of

Comment (micahriggan) [reply lib/btc/BtcRpc.js]: If I pass an array into this method, does this mutate my array?

Comment (sbhatnagar30) [reply lib/btc/BtcRpc.js]: you pass in an array of objects and get back the same array of objects but with additional params in each object

File: lib/btc/BtcRpc.js [hunk 11]
```diff
@@ -53,65 +65,83 @@ class BtcRpc {
...
+        }
+        let tx = payToArrayClone.shift();
+        tempArray.push(tx);
+        let address = tx.address;
+        let amount = tx.amount;
+        if (batch[address]) {
+          batch[address] = 0;
+        }
+        batch[address] += amount;
+        batch[address] = amount;
```
Comment (micahriggan) [inline lib/btc/BtcRpc.js]: line should be removed.

**Pull request merged**